sql/codegen_alps.go

```diff
@@ -65,346 +51,13 @@ type alpsFiller struct {
-		cc.Key, cc.BucketSize), nil
-}
-
-func (ec *embeddingColumn) GenerateCode() (string, error) {
-	catColumn, ok := ec.CatColumn.(featureColumn)
-	if !ok {
-		return "", fmt.Errorf("embedding generate code error, input is not featureColumn: %s", ec.CatColumn)
-	}
-	sourceCode, err := catColumn.GenerateCode()
-	if err != nil {
-		return "", err
-	}
-	return fmt.Sprintf("tf.feature_column.embedding_column(%s, dimension=%d, combiner=\"%s\")",
-		sourceCode, ec.Dimension, ec.Combiner), nil
-}
-
-func (a *attribute) GenerateCode() (string, error) {
-	if val, ok := a.Value.(string); ok {
-		return fmt.Sprintf("%s=\"%s\"", a.Name, val), nil
-	}
-	if val, ok := a.Value.([]interface{}); ok {
-		intList, err := transformToIntList(val)
-		if err != nil {
-			return "", err
-		}
-		return fmt.Sprintf("%s=%s", a.Name,
-			strings.Join(strings.Split(fmt.Sprint(intList), " "), ",")), nil
-	}
-	return "", fmt.Errorf("value of attribute must be string or list of int, given %s", a.Value)
-}
-
-func (fs *featureSpec) ToString() string {
-	if fs.IsSparse {
-		return fmt.Sprintf("SparseColumn(name=\"%s\", shape=%s, dtype=\"%s\", separator=\"%s\")",
-			fs.FeatureName,
-			strings.Join(strings.Split(fmt.Sprint(fs.Shape), " "), ","),
-			fs.DType,
-			fs.Delimiter)
-	}
-	return fmt.Sprintf("DenseColumn(name=\"%s\", shape=%s, dtype=\"%s\", separator=\"%s\")",
-		fs.FeatureName,
-		strings.Join(strings.Split(fmt.Sprint(fs.Shape), " "), ","),
-		fs.DType,
-		fs.Delimiter)
-}
-
-func resolveFeatureSpec(el *exprlist, isSparse bool) (*featureSpec, error) {
-	if len(*el) != 4 {
-		return nil, fmt.Errorf("bad FeatureSpec expression format: %s", *el)
-	}
-	name, err := expression2string((*el)[1])
-	if err != nil {
-		return nil, fmt.Errorf("bad FeatureSpec name: %s, err: %s", (*el)[1], err)
-	}
-	shape, err := strconv.Atoi((*el)[2].val)
-	if err != nil {
-		return nil, fmt.Errorf("bad FeatureSpec shape: %s, err: %s", (*el)[2].val, err)
-	}
-	unresolvedDelimiter, err := expression2string((*el)[3])
-	if err != nil {
-		return nil, fmt.Errorf("bad FeatureSpec delimiter: %s, err: %s", (*el)[1], err)
-	}
-
-	delimiter, err := resolveDelimiter(unresolvedDelimiter)
-	if err != nil {
-		return nil, err
-	}
-	// TODO(uuleon): hard coded dtype(float) should be removed
-	return &featureSpec{
-		FeatureName: name,
-		IsSparse:    isSparse,
-		Shape:       []int{shape},
-		DType:       "float",
-		Delimiter:   delimiter}, nil
-}
-
-func resolveExpression(e interface{}) (interface{}, error) {
-	if expr, ok := e.(*expr); ok {
-		if expr.val != "" {
-			return expr.val, nil
-		}
-		return resolveExpression(&expr.sexp)
-	}
-
-	el, ok := e.(*exprlist)
-	if !ok {
-		return nil, fmt.Errorf("input of resolveExpression must be `expr` or `exprlist`: %s", e)
-	}
-
-	headName := (*el)[0].val
-	if headName == "" {
-		return resolveExpression(&(*el)[0].sexp)
-	}
-
-	headName = strings.ToUpper(headName)
-
-	switch headName {
-	case dense:
```

Comment:
I think `DENSE` and `SPARSE` only work for ALPS. You can refer to the generated code at https://github.com/sql-machine-learning/sqlflow/pull/502#issuecomment-501077347 where `DENSE` and `SPARSE` are translated to `DenseColumn` and `SparseColumn` in `DataX`, instead of `tf.feature_column.*` in the model constructor.

@uuleon I am not sure if ALPS has a plan for releasing these `DENSE` and `SPARSE` as part of the extension for feature column.

Reply:
Thanks for reminding, I am moving on to make this resolver work for both submitters in later PRs.